experimental/devices/ina219/ina219.go

```diff
@@ -0,0 +1,280 @@
+	// One rx buffer for entire transaction
+	rx := make([]byte, 2)
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if err := d.c.Tx([]byte{shuntVoltageRegister}, rx); err != nil {
+		return PowerMonitor{}, errReadShunt
+	}
+	var pm PowerMonitor
+	pm.Shunt = physic.ElectricPotential(int16(binary.BigEndian.Uint16(rx)))
+	// Least significant bit is 10µV
+	pm.Shunt *= (10 * physic.MicroVolt)
+
+	if err := d.c.Tx([]byte{busVoltageRegister}, rx); err != nil {
+		return PowerMonitor{}, errReadBus
+	}
+	// check bit 0 of bus voltage register, if set data is invalid
+	if hasBit(rx[1], 0) {
+		return PowerMonitor{}, errRegisterOverflow
+	}
+	pm.Voltage = physic.ElectricPotential(int16(binary.BigEndian.Uint16(rx) >> 3))
+	// Least significant bit is 4mV
+	pm.Voltage *= (4 * physic.MilliVolt)
+
+	// if calibration register is not set then current and power readings are
+	// meaningless
+	if d.caibrated {
+		if err := d.c.Tx([]byte{currentRegister}, rx); err != nil {
+			return PowerMonitor{}, errReadCurrent
+		}
+		pm.Current = physic.ElectricCurrent(binary.BigEndian.Uint16(rx))
+		pm.Current *= d.currentLSB
+
+		if err := d.c.Tx([]byte{powerRegister}, rx); err != nil {
+			return PowerMonitor{}, errReadPower
+		}
+		pm.Power = physic.Power(binary.BigEndian.Uint16(rx))
+		pm.Power *= d.powerLSB
+	}
+
+	return pm, nil
+}
+
+// Since physic electrical is in nano units we need
+// to scale taking care to not overflow int64 or loose resolution.
+const calibratescale int64 = ((int64(physic.Ampere) * int64(physic.Ohm)) / 100000) << 12
+
+// Calibrate sets the scaling factor of the current and power registers for the
+// maximum resolution. Calibrate is run on init. here it allows you to make
+// tune the measured value with actual value
+func (d *Ina219) Calibrate() error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.sense < 1 {
+		return errSenseResistorValueInvalid
+	}
+	if d.maxCurrent < 1 {
+		return errMaxCurrentInvalid
+	}
+	d.currentLSB = d.maxCurrent / (2 << 15)
+	d.powerLSB = physic.Power(d.currentLSB * 20)
+	// cal = 0.04096 / (current LSB * Shunt Resistance) where lsb is in Amps and
+	// resistance is in ohms.
+	// calibration register is 16 bits wide.
+	cal := uint16(calibratescale / (int64(d.currentLSB) * int64(d.sense)))
+	if err := d.WriteRegister(calibrationRegister, cal); err != nil {
+		return err
+	}
+
+	d.caibrated = true
+	return nil
+}
+
+// Reset performs a power on reset.
+func (d *Ina219) reset() error {
+	// reset Ina219ice by writing 1 to the 15th bit in register 0x00
+	if err := d.c.Tx([]byte{0x00, 0x80, 0x00}, nil); err != nil {
+		return err
+	}
+	// give ina219 time to preform reset.
+	time.After(time.Millisecond * 20)
+	rx := make([]byte, 2)
+	if err := d.c.Tx([]byte{0x00}, rx); err != nil {
+		return err
+	}
+
+	const powerOnReset = 0x399F // register 0x00 on POR
+	if binary.BigEndian.Uint16(rx) != powerOnReset {
+		return errResetError
+	}
+	return nil
+}
+
+// WriteRegister is used to write to register address directly
```

Comment:
Are you sure you want to expose this function?

Also note that https://periph.io/x/periph/conn/mmr#Dev8 may be helpful here. Not required, just FYI.

Reply:
not really, I misread a driver requirement. 

Had not seen mmr, it clean up some of that code nicely.